pep-0716.rst

```diff
@@ -0,0 +1,319 @@
+   specified versions.
+
+   This normalization used ``-``, which was an invalid value for :pep:`427` and
+   required escaping to ``_``, so :pep:`440` was extended to allow that as an
+   optional spelling of ``_``, which would normalize to ``-``.
+3. :pep:`503` was accepted, which specified a normalization of the project name
+   *when* querying the Simple API for a project.
+4. The spec for ``.dist-info`` required normalization of the name (using the
+   :pep:`503` rules) but did not specify a requirement on version. The :pep:`503`
+   normalization uses ``-``, but tools that locate ``.dist-info`` use the ``-``
+   character to split between name and version, so in practice nobody was
+   following this requirement.
+
+   Thus, the ``.dist-info`` spec was `updated <https://github.com/pypa/packaging.python.org/pull/781>`__,
+   without a PEP, to make the spec more closely align with common practice. The
+   result of that being that the spec states that name must be normalized as
+   per :pep:`503` and versions must be normalized as per :pep:`440`, but
+   escaping ``-`` with ``_``.
+
+   This change recognized that there are many existing ``.dist-info`` directories
+   that are not normalized, and thus instructs tools to expect ``.dist-info``
+   directories with unnormalized values, but that all tools must write normalized
+   values going forward.
+5. It was `noted <https://discuss.python.org/t/5605>`__
+   that :pep:`427` required that the segments of the filename contain only
+   alphanumeric characters, ``_``, and ``.``, and that all other characters must
+   be escaped with ``_``. However, :pep:`440` allows the use of ``!`` and ``+``,
+   which meant that those characters got escaped to ``_``, which could then not
+   be parsed back into their original versions.
+6. As a result of that discussion, the Wheel specs were `updated <https://discuss.python.org/t/5605>`__
+   and then `updated again <https://github.com/pypa/packaging.python.org/pull/1032>`__,
+   without a PEP, to require that versions were normalized using :pep:`440` and
+   then ``-`` was escaped with ``_``.
+
+   That change also required that runs of ``-_.`` should be replaced with ``_``
+   as well as lowercasing everything. It noted that it was equivalent to :pep:`503`
+   normalization followed by replacing ``-`` with ``_``.
+
+   It was `noted 9 months later <https://discuss.python.org/t/5605/21>`__, that
+   there wasn't much discussion on the change to name normalization, but that it
+   landed anyways.
+7. :pep:`621` was accepted, which provided a way to specify project metadata in
+   ``pyproject.toml``. This PEP was careful to make the distinction between
+   static metadata, where tools could trust the values in ``pyproject.toml`` and
+   dynamic metadata where they could not.
+
+   However, this PEP doesn't clarify whether static values must be identical
+   values or equivalent values. To make matters worse, it includes the statement
```

Comment:
I disagree. I think it simply wasn't considered. At the time, people were *definitely* talking about reading metadata from `pyproject.toml` if it was specified as static. That's only reasonable if there's a presumption of the data being copied exactly. In reality, the notion of reading from `pyproject.toml` was considered "possible, but not recommended" - which led directly to me writing PEP 643 (Metadata for Source Distributions). But I don't recall anyone at the time discussing the issue in enough detail to pick up on the question of transformations.

Reply:
Sorry for the confusion, I didn't mean to speak to what people thought at the time the PEP was written here (as I wasn't around for that, unlike yourself—thanks for the insight!). And I definitely agree that `Name` MUST NOT be normalized in the `METADATA`.

To clarify the above, what I meant was that ultimately, what seems to have been settled on is that the output `METADATA`, minus pyproject metadata keys marked dynamic, can be statically and deterministically extracted from a given `pyproject.toml` following only what is specified in the current "Declaring project metadata" spec. However, its hard to see how people could have previously interpreted it as an identical 1:1 copy, as the `[project]` table keys do not map 1:1 to `METADATA` fields, and the original PEP itself already required certain non-trivial transformations, e.g. for `author`/`author-email` and `maintainer`/`maintainer-email`, among others. And of course, Brett's own PEP 685 followup requires normalization for extras names in `[project]`.

By contrast, as you mention, AFAIK PEP 643 _does_ entail that non-dynamic sdist `PKG-INFO` fields can be copied 1:1, since there is an exact 1:1 mapping there any transformations from the user-entered data that will end up in the `METADATA` should have already been applied in the `PKG-INFO`.